learn/best_practices/type_casting.md

```diff
@@ -0,0 +1,494 @@
+
+Object Oriented Approach
+------------------------
+
+The module:
+
+``` fortran
+module integrals
+
+  use types, only: dp
+  implicit none
+  private
+
+  public :: integrand, simpson
+
+  ! User extends this type
+  type, abstract :: integrand
+  contains
+    procedure(func), deferred :: eval
+  end type
+
+  abstract interface
+    function func(this, x) result(fx)
+      import :: integrand, dp
+      class(integrand) :: this
+      real(dp), intent(in) :: x
+      real(dp) :: fx
+    end function
+  end interface
+
+contains
+
+real(dp) function simpson(f, a, b) result(s)
+  class(integrand) :: f
+  real(dp), intent(in) :: a, b
+  s = ((b-a)/6) * (f%eval(a) + 4*f%eval((a+b)/2) + f%eval(b))
+end function
+
+end module
+```
+
+The abstract type prescribes exactly what the integration routine needs,
+namely a method to evaluate the function, but imposes nothing else on
+the user. The user extends this type, providing a concrete
+implementation of the eval type bound procedure and adding necessary
+context data as components of the extended type.
+
+Usage:
+
+``` fortran
+module example_usage
+
+  use types, only: dp
+  use integrals, only: integrand, simpson
+  implicit none
+  private
+
+  public :: foo
+
+  type, extends(integrand) :: my_integrand
+    real(dp) :: a, k
+  contains
+    procedure :: eval => f
+  end type
+
+contains
+
+function f(this, x) result(fx)
+  class(my_integrand) :: this
+  real(dp), intent(in) :: x
+  real(dp) :: fx
+  fx = this%a*sin(this%k*x)
+end function
+
+subroutine foo(a, k)
+  real(dp) :: a, k
+  type(my_integrand) :: my_f
+  my_f%a = a
+  my_f%k = k
+  print *, simpson(my_f, 0.0_dp, 1.0_dp)
+  print *, simpson(my_f, 0.0_dp, 2.0_dp)
+end subroutine
+
+end module
+```
+
+Complete Example of void \* vs type(c\_ptr) and transfer()
+----------------------------------------------------------
+
+Here are three equivalent codes: one in C using `void *` and two codes
+in Fortran using `type(c_ptr)` and `transfer()`:
+
+| Language &nbsp; | Method               | Link                              |
+|-----------------|----------------------|-----------------------------------|
+| C               | `void *`             | <https://gist.github.com/1665641> |
+| Fortran         | `type(c_ptr)` &nbsp; | <https://gist.github.com/1665626> |
+| Fortran         | `transfer()`         | <https://gist.github.com/1665630> |
+
+The C code uses the standard C approach for writing extensible libraries
+that accept callbacks and contexts. The two Fortran codes show how to do
+the same. The `type(c_ptr)` method is equivalent to the C version and
+that is the approach that should be used.
+
+The `transfer()` method is here for completeness only (before Fortran
+2003, it was the only way) and it is a little cumbersome, because the
+user needs to create auxiliary conversion functions for each of his
+types. As such, the `type(c_ptr)` method should be used instead.
```

Comment:
Can't we just move the code snippets here? Linking against gist doesn't really look nice IMO.